### PR TITLE
Material Cockpit performance improvements

### PR DESCRIPTION
**Pull request by mrie04**

### CODE REVIEW

File: backend/de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/warehouse/WarehouseRepository.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+{
+    IQueryBL queryBL = Services.get(IQueryBL.class);
+    @VisibleForTesting
+    public static WarehouseRepository newInstanceForUnitTesting()
+    {
+        Adempiere.assertUnitTestMode();
+        return new WarehouseRepository();
+    }
+
+    private final CCache<WarehouseId, Warehouse> cache = CCache.<WarehouseId, Warehouse>builder()
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/warehouse/WarehouseRepository.java:49]: why caching per ID and not caching all?


also, what about WarehouseDAO?

File: backend/de.metas.business/src/main/java/de/metas/product/ProductRepository.java [hunk 2]
```diff
@@ -60,6 +65,44 @@ public class ProductRepository
 {
 	private final IQueryBL queryBL = Services.get(IQueryBL.class);
 	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	private final IUOMDAO uomDAO = Services.get(IUOMDAO.class);
+
+	private final CCache<ProductId, CockpitProduct> cache = CCache.<ProductId, CockpitProduct>builder()
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/product/ProductRepository.java:70]: why is this called CockpitProduct in such a generic service?

File: backend/de.metas.business/src/main/java/de/metas/product/ProductRepository.java [hunk 3]
```diff
@@ -60,6 +65,44 @@ public class ProductRepository
...
+			.build();
+
+	@VisibleForTesting
+	public static ProductRepository newInstanceForUnitTesting()
+	{
+		Adempiere.assertUnitTestMode();
+		return new ProductRepository();
+	}
+
+	public CockpitProduct getCockpitProductById(@NonNull final ProductId productId)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/product/ProductRepository.java:83]: why here?

File: backend/de.metas.business/src/main/java/de/metas/product/ProductRepository.java [hunk 4]
```diff
@@ -60,6 +65,44 @@ public class ProductRepository
...
+	{
+		return cache.getOrLoad(productId, this::retrieveCockpitProductById);
+	}
+
+	private CockpitProduct retrieveCockpitProductById(@NonNull final ProductId productId)
+	{
+		return cockpitProductFromRecord(loadOutOfTrx(productId, I_M_Product.class));
+	}
+
+	private CockpitProduct cockpitProductFromRecord(@NonNull final I_M_Product productRecord)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/product/ProductRepository.java:93]: static?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowCache.java [hunk 5]
```diff
@@ -22,8 +27,9 @@
 @RequiredArgsConstructor
 public class MaterialCockpitRowCache
 {
-	private final ProductRepository productRepository;
-	private final IUOMDAO uomDAO;
+	@NonNull private final ProductRepository productRepository = SpringContextHolder.instance.getBean(ProductRepository.class);
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowCache.java:28]: pls avoid getBean here. Provide those as ctor params.

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRow.java [hunk 6]
```diff
@@ -412,27 +408,24 @@ private MaterialCockpitRow(
...
-		final IProductDAO productDAO = Services.get(IProductDAO.class);
-
-		final I_M_Product productRecord = productDAO.getById(productId);
-		final I_M_Product_Category productCategoryRecord = productDAO.getProductCategoryById(ProductCategoryId.ofRepoId(productRecord.getM_Product_Category_ID()));
+		final Product cockpitProduct = cache.getProductById(productId);
 
-		this.productValue = productRecord.getValue();
-		this.productName = productRecord.getName();
+		this.productValue = cockpitProduct.getValue();
+		this.productName = cockpitProduct.getName().getDefaultValue();
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRow.java:414]: afair, productName can be ITranslatableString.

**Review by teosarca — CHANGES_REQUESTED** [review]
check my cmments

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowCache.java (hunk 5 above)
Comment (mrie04) [reply backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowCache.java:28]: Then I run into may not be initialized error and I couldn't figure out why

File: backend/de.metas.business/src/test/java/de/metas/business/BusinessTestHelper.java [hunk 7]
```diff
@@ -209,6 +209,7 @@ public I_M_Product createProduct(
 		product.setC_UOM_ID(UomId.toRepoId(uomId));
 		product.setProductType(ProductType.Item.getCode());
 		product.setIsStocked(true);
+		product.setM_Product_Category_ID(ProductCategoryId.toRepoId(createProductCategory("Standard", null)));
```
Comment (teosarca) [inline backend/de.metas.business/src/test/java/de/metas/business/BusinessTestHelper.java:212]: so u will create a new product category each time you will create a product?

Comment (mrie04) [reply backend/de.metas.business/src/test/java/de/metas/business/BusinessTestHelper.java]: adjusted

File: backend/de.metas.business/src/test/java/de/metas/business/BusinessTestHelper.java [hunk 8]
```diff
@@ -230,6 +231,16 @@ public ProductCategoryId createProductCategory(@NonNull final String name, @Null
 		return ProductCategoryId.ofRepoId(category.getM_Product_Category_ID());
 	}
 
+	public ProductCategoryId createStandardProductCategory()
+	{
+		final I_M_Product_Category category = newInstance(I_M_Product_Category.class);
+		category.setM_Product_Category_ID(ProductCategoryId.toRepoId(defaultProductCategoryId));
```
Comment (teosarca) [inline backend/de.metas.business/src/test/java/de/metas/business/BusinessTestHelper.java:237]: pls check if alrady exists first...

Comment (mrie04) [reply backend/de.metas.business/src/test/java/de/metas/business/BusinessTestHelper.java]: done

**Review by teosarca — APPROVED** [review]

**Pull request merged**